File: Data Structures & Algorithms/search-2d-matrix/submission-3.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {

        for (int row = 0; row < matrix.length; row++) {
            int l = 0, r = matrix[0].length - 1;
            while (l <= r) {
                int m = (l + r) / 2;
                if (matrix[row][m] < target) {
                    l++;
                } else if (matrix[row][m] > target) {
                    r--;
                } else {
                    return true;
                }
            }
        } 

        return false;   
    }
}

File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-3.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        N = len(grid) 
        if grid[0][0] == 1 or grid[N - 1][N - 1] == 1:
            return -1
        
        q = deque([(0, 0, 1)])      # (row, col, length)
        directions = [[-1, 0], [0, 1], [1, 0], [0, -1],
                      [-1, -1], [1, -1], [1, 1], [-1, 1]]
        visit = set([(0, 0)])     # (row, col)

        while q:
            row, col, length = q.popleft()
            if row == N - 1 and col == N - 1:
                return length 
            for dr, dc in directions:
                nr, nc = dr + row, dc + col 
                if (0 <= nr < N and 0 <= nc < N and
                    (nr, nc) not in visit and grid[nr][nc] == 0):
                    visit.add((nr, nc))
                    q.append((nr, nc, length + 1))
        return -1